manual/manual/html_processing/src/process_api.ml

```diff
@@ -0,0 +1,414 @@
+   forces to click twice to an external link after entering text. *)
+let search_widget with_description =
+  sprintf "<div class=\"api_search\"><input type=\"text\" name=\"apisearch\" id=\"api_search\"
+	 oninput    = \"mySearch(%b);\"
+         onkeypress = \"this.oninput();\"
+         onclick    = \"this.oninput();\"
+	 onpaste    = \"this.oninput();\">
+<img src=\"search_icon.svg\" alt=\"Search\" class=\"svg\" onclick=\"mySearch(%b)\">%s</div>
+<div id=\"search_results\"></div>" with_description with_description
+    (if with_description
+     then {|<span class="search_comment">(search values, type signatures, and descriptions - case sensitive)<div class="search_help"><ul><li>You may search bare values, like <code>map</code>, or indicate the module, like <code>List.map</code>, or type signatures, like <code>int -> float</code>.</li><li>To combine several keywords, just separate them by a space - except if you want to search type signatures: then you must use <strong>two</strong> spaces as separators.</li><li>You may use the special chars <code>^</code> and <code>$</code> to indicate where the matched string should start or end, respectively.</li></ul></div></span>|}
+     else "")
+  |> parse
+
+let process ?(search=true) ~version file out =
+
+  sprintf "Processing %s..." file |> pr;
+  let html = read_file file in
+  let soup = parse html in
+
+  (* Add javascript files *)
+  if search then begin
+    let head = soup $ "head" in
+    create_element "script" ~attributes:["src","search.js"]
+    |> append_child head;
+    create_element "script" ~attributes:["src","scroll.js"]
+    |> append_child head;
+  end;
+
+  (* Add favicon *)
+  let head = soup $ "head" in
+  {|<link rel="shortcut icon" type="image/x-icon" href="|} ^ favicon () ^ {|">|}
+  |> parse
+  |> append_child head;
+
+  (* Add api wrapper *)
+  let body = wrap_body ~classes:["api"] soup in
+
+  (* Delete previous/up/next links *)
+  body $? "div.navbar"
+  |> do_option delete;
+
+  (* Create TOC with H2 and H3 elements *)
+  (* It would be much simpler to insert only H2 *)
+  let header = create_element "header" in
+  prepend_child body header;
+  let nav = create_element "nav" ~class_:"toc" in
+  append_child header nav;
+  let ul = create_element "ul" in
+  append_child nav ul;
+  let h3_open = ref false in
+  let h3_current = ref (create_element "ul") in
+  let li_current = ref (create_element "li") in
+  (* Create a "li" element inside "ul" from a header "h" (h2 or h3 typically) *)
+  let li_of_h ul h =
+    li_current := create_element "li";
+    append_child ul !li_current;
+    match attribute "id" h with
+    | Some id ->
+        let href = "#" ^ id in
+        let a = create_element "a" ~inner_text:(texts h |> String.concat "") ~attributes:["href", href] in
+        append_child !li_current a
+    | None -> () in
+
+  descendants body
+  |> elements
+  |> iter (fun h -> match name h with
```

Comment:
Can't you use `fold` here instead, and have h3_open as the value you're folding?

Reply:
OK, then I can even fold over `li_current,h3_current` and remove `h3_open`